redbox-core/redbox/loader/loaders.py

```diff
@@ -135,77 +62,30 @@ def _chunking(self) -> Any:
-            self.metadata = self.default_metadata
-
-        else:
-            # Get first 1k tokens of processed document
-            first_n = self.get_first_n_tokens(elements, 1_000)
+        chunks = self._chunking()
+        first_thousand_words = "".join(chunk["text"] for chunk in chunks)[:1_000]
+        metadata = self.create_file_metadata(first_thousand_words)
+        return GeneratedMetadata.model_validate(metadata).model_dump()
 
-            # Get whatever metadata we can from processed document
-            doc_metadata = self.get_doc_metadata(chunks=elements, n=3, ignore=None, max_size=524288 - len(first_n))
-
-            # Generate new metadata
-            res = self.create_file_metadata(first_n, doc_metadata)
-            if not res:
-                self.metadata = self.default_metadata
-                logger.warning("LLM failed to extract metadata for this file")
-            else:
-                if all(key in res.keys() for key in self.required_keys):
-                    self.metadata = res
-                else:
-                    # missing keys
-                    self.metadata = self.default_metadata
-
-    def create_file_metadata(self, page_content: str, metadata: dict[str, Any]) -> dict[str, Any]:
+    def create_file_metadata(self, page_content: str) -> GeneratedMetadata:
         """Uses a sample of the document and any extracted metadata to generate further metadata."""
-        metadata_chain = (
-            ChatPromptTemplate.from_messages(
-                [
-                    self.env.metadata_prompt,
-                    (
-                        "user",
-                        (
-                            "<metadata>\n"
-                            "{metadata}\n"
-                            "</metadata>\n\n"
-                            "<document_sample>\n"
-                            "{page_content}"
-                            "</document_sample>"
-                        ),
-                    ),
-                ]
-            )
-            | self.llm
-            | JsonOutputParser()
-        )
-
-        try:
-            return metadata_chain.invoke({"page_content": page_content, "metadata": metadata})
-        except ConnectionError as e:
-            logger.warning(f"Retrying due to HTTPError {e}")
-        except json.JSONDecodeError:
-            # replace with fail safe metadata
-            return None
-        except Exception as e:
-            raise Exception(f"Some error happened in metadata extraction. {e}")
 
+        parser = PydanticOutputParser(pydantic_object=GeneratedMetadata)
+        metadata_prompt = PromptTemplate(
+            template="".join(self.env.metadata_prompt) + "\n\n{format_instructions}\n\n{page_content}\n",
+            input_variables=["page_content"],
+            partial_variables={"format_instructions": parser.get_format_instructions()},
+        )
+        metadata_chain = metadata_prompt | self.llm | JsonOutputParser()
```

Comment:
shouldn't we use the PydanticOutputParser from above here?

Reply:
yes, good point!

Reply:
[9873ba2](https://github.com/i-dot-ai/redbox/pull/1147/commits/9873ba2a0af6bbdf331e59330bae6b1c389dbf7d)